File: 0438-find-all-anagrams-in-a-string/0438-find-all-anagrams-in-a-string.java
class Solution {
    public List<Integer> findAnagrams(String s, String p) {
        List<Integer> result = new ArrayList<>();
        if(s == null || p == null || s.length() < p.length()){
            return result;
        }
        
        int sCount[] = new int[26];
        int pCount[] = new int[26];
        
        int k = p.length();
        for(int i=0; i<k; i++){
            pCount[p.charAt(i) - 'a']++;
        }
        
        for(int i=0; i<k; i++){
            sCount[s.charAt(i) - 'a']++;
        }
        
        if(Arrays.equals(sCount, pCount)){
            result.add(0);
        }
        
        int left = 0;
        for(int right=k; right<s.length(); right++){
            sCount[s.charAt(left) - 'a']--;
            left++;
            
            sCount[s.charAt(right) - 'a']++;
            
            if(Arrays.equals(sCount, pCount)){
                result.add(left);
            }
        }
        return result;
    }
}

File: 0438-find-all-anagrams-in-a-string/0438-find-all-anagrams-in-a-string.cpp
class Solution {
public:
    vector<int> findAnagrams(string s, string p) {
        vector<int> result;
        if(s.empty() || p.empty() || s.size() < p.size()){
            return result;
        }
        
        vector<int> sCount(26, 0);
        vector<int> pCount(26, 0);
        
        for(char ch: p){
            pCount[ch - 'a']++;
        }
        
        int k = p.size();
        for(int i=0; i<k; i++){
            sCount[s[i] - 'a']++;
        }
        
        if(sCount == pCount){
            result.push_back(0);
        }
        
        int left = 0;
        for(int right=k; right<s.size(); right++){
            sCount[s[left] - 'a']--;
            left++;
            
            sCount[s[right] - 'a']++;
            
            if(sCount == pCount){
                result.push_back(left);
            }
        }
        return result;
    }
};